src/Database/TigerBeetle/Raw/Client.hs

```diff
@@ -146,3 +145,110 @@ withClient cfg clusterId address action =
+            (finalizeClient clientState)
+        _ -> pure $ Left initStatus
+
+finalizeClient :: ClientState -> IO ()
+finalizeClient state = do
+  -- Signal that no new incoming request should proceed
+  atomically $ writeTVar state.csIsShutdown True
+
+  -- N.B. race-condition: Once we release from this transaction
+  -- completionCallbacks may fire and overwrite our shutdown error in the
+  -- request resultVar. Conceivably this is fine, the below block just ensures
+  -- that all submitRequest invocations that are blocked on the request context
+  -- resultVar are allowed to proceed. This should effectively flush all active
+  -- requests since the isShutdown flag should prevent new requests from coming
+  -- in.
+  --
+  -- Process interleaving shouldn't cause any active request entries after
+  -- this block (due to using a stale isShutdown value) because the flag is
+  -- read in the same atomically block as active requests are written to in
+  -- submitRequest function
+  atomically $ do
+    activeReqs <- readTVar state.csActiveRequests 
+    -- Iterate through all the request vars and put a client shutdown result 
+    -- so that instances of `submitRequest` that are waiting are unblocked
+    forM_ (IM.elems activeReqs) $ \context ->
+      putTMVar context.resultVar (Left ClientShutdownDuringRequest)
+  
+  -- De-initialize the client
+  void $ tbClientDeinit state.csClientPtr
+
+submitRequest 
+  :: ClientState
+  -> TBOperation
+  -> ByteString  -- ^ Request data
+  -> IO (Either RequestError TBResponse)
+submitRequest state operation reqData = do
+  -- Scaffold request state
+  res <- atomically $ readTVar state.csIsShutdown >>= \case
+    True -> pure (Left ClientShutdownDuringRequest)
+    False -> Right <$> provisionRequestContext state
+
+  case res of
+    Left e -> pure $ Left e 
+    Right context -> withPacketPtrs reqData \packetPtr contextPtr (dataPtr, dataSize) -> do
+      -- TODO: confirm that this is the correct way to assign a context id to a packet
+      poke contextPtr context.contextId
+      let packet = TBPacket
+            { tbPacketUserData = castPtr contextPtr
+            , tbPacketData = castPtr dataPtr
+            , tbPacketDataSize = fromIntegral dataSize
+            , tbPacketUserTag = 0
+            , tbPacketOperation = operation
+            , tbPacketStatus = Ok
+            , tbPacketOpaque = V.empty
+            }
+      poke packetPtr packet
+
+      -- Submit the request
+      submitStatus <- tbClientSubmit state.csClientPtr packetPtr
+
+      case submitStatus of
+        ClientOk -> do
+          -- Wait for the result with a timeout
+          let timeoutMicros = fromIntegral state.csTimeoutMillis * 1000  -- Convert ms to μs
+          result <- timeout timeoutMicros $ atomically $ takeTMVar context.resultVar            
+          case result of
+            Just r -> pure r
+            Nothing -> cleanupRequest state context $> Left (RequestTimeoutError operation reqData)
```

Comment:
We may want a higher level interface that provides the user with something they can explicitly await (maybe TVar, or maybe custom api we come up with). Another approach might be to give the user a utility that runs a `submitRequest` on a separate thread, that they can check in on as needed.